### ISSUE TITLE
[feature] Add snippets / ability to add snippets to quicker launch apps from the search bar.

### ISSUE THREAD

**Issue by 5ven5chuette** [jrpie/Launcher#88]
Snippets instead of first unique name match in search bar

Example: 

Instead of having to type until the launcher recognizes a unique name, make it so I can just type "gh" for GitHub or "dc" for Discord.

**Comment by jrpie** [comment]
Thanks for the suggestion!

You can already get a very similar behavior by renaming apps, e.g. rename github to something like "github (gc)" or "github - gh"

**Comment by wassupluke** [comment]
Dude I fricken love this idea. It's like a bash alias. Game changer. 🙏🏻🙌🏻

**Comment by jrpie** [comment]
> Though, depending on the alias you rename it to, the search may still find other app matches. i.e., making sure someone can use an alias "gm" for some app doesn't preclude them from then being able to get to their Gmail app with the same first letter inputs of "gm" would be wise. I could see people getting frustrated by that.

I'm not really sure how to solve this. For shortcuts to be useful they should be launched immediately (otherwise this feature would not add much over just renaming the app). On the other hand it might make the usual search really frustrating.

**Comment by 5ven5chuette** [comment]
> Thanks for the suggestion!
> 
> You can already get a very similar behavior by renaming apps, e.g. rename github to something like "github (gc)" or "github - gh" 

Thanks for pointing this out, haven't noticed, yet.

Theoretically this makes my request obsolete, practically it would still be neat to provide a functionality to add a snippet that gets stored in the background while the application name on the list still is verbose.

So when I type gh it immidiately opens GitHub while the app name remains original.

**Comment by jrpie** [comment]
Yes, so snippets have two benefits over renaming:
* displayed names don't change
* launch immediately even if there are other names matching the search (not sure whether this is a feature or an anti-feature ^^ one might argue that choosing sane snippet names is the responsibility of the user)

I'm not really sure whether that is worth the effort (implementing UI for managing snippets etc.) and added complexity.

**Comment by jrpie** [comment]
Somewhat unrelated: Thinking about this, I had the idea to let users launch a web search from the search list: Currently pressing the submit button on the keyboard launches the first matching app. Adding an option to launch a web search instead would be relatively straight forward.
(Doesn't work too well with auto-launch though)

**Comment by wassupluke** [comment]
Would it use a static search engine or a user preference engine? I'm thinking it could allow the user to pass "g cat pics" to use Google engine to search cat pics, or something like "d " or "b " for duckduckgo and bing, respectively. I think the version of Internet Explorer my old job used allowed this address bar functionality where "g search query" would let you Google for a search query (without the g it would try to resolve the input as a path on the LAN).

**Comment by 5ven5chuette** [comment]
> Yes, so snippets have two benefits over renaming:
> * displayed names don't change
> * launch immediately even if there are other names matching the search (not sure whether this is a feature or an anti-feature ^^ one might argue that choosing sane snippet names is the responsibility of the user)
> 
> I'm not really sure whether that is worth the effort (implementing UI for managing snippets etc.) and added complexity.

Well, yeah, there would be a responsibility given to the user, unless (but that adds even more work regarding implementation I guess) you restrict snippet names and make it so there can't be multiples of the same abbreviation. Maybe even restrict the length of the given name, since those are only snippets.

**Comment by 5ven5chuette** [comment]
> Somewhat unrelated: Thinking about this, I had the idea to let users launch a web search from the search list: Currently pressing the submit button on the keyboard launches the first matching app. Adding an option to launch a web search instead would be relatively straight forward. (Doesn't work too well with auto-launch though)

Doesn't the first matching app gets launched directly as soon as the search bar identifies a unique name? At least thats the case with me.

**Comment by wassupluke** [comment]
> the first matching app gets launched directly

Correct, if "Launch search results" is toggled on.

**Comment by jrpie** [comment]
> Would it use a static search engine or a user preference engine?

The easiest way would be to use 
[Intent.ACTION_WEB_SEARCH](https://developer.android.com/reference/android/content/Intent#ACTION_WEB_SEARCH).

Alternatively I could open a (user provided) URL like ` https://www.startpage.com/search?query=%s&language=auto`

**Comment by wassupluke** [comment]
I'd say KISS for now with the intent and if enough users ask for the ability to use their own then you could tweak it down the road.

**Comment by jrpie** [comment]
The auto-launch problem was already fixed in 0.0.20: you can prefix the search with space to suppress auto launch (this was suggested to me by mail, I really should document that somewhere)
I think space is the optimal solution to this problem as it is very easy to reach and usually not needed as the first character in a search query.

**Comment by jrpie** [comment]
> Oh okay, sorry!

Nothing to be sorry about, as I said, I really should document this somewhere (cf. #91)

**Comment by jrpie** [comment]
> Regarding custom search engines I think I prefer simply using [Intent.ACTION_WEB_SEARCH](https://developer.android.com/reference/android/content/Intent#ACTION_WEB_SEARCH) (that is what is currently implemented), since [...]


I built a small proof of concept ([code](https://git.jrpie.de/jrpie/usearch), [debug build](https://jrpie.de/downloads/usearch-0.0.1.apk))

**Comment by wassupluke** [comment]
> you can prefix the search with space to suppress auto launch

Yeah, didn't know that feature came out. Super slick integration!!!!! 👌🏻

**Comment by jrpie** [comment]
Thinking about this some more, the entire snippets feature could be implemented by a companion app like the [µSearch PoC](https://git.jrpie.de/jrpie/usearch) I built yesterday.
That way, it could be installed optionally, wouldn't get in the way for users who don't need it and would be very easy to replace by custom logic. A potential downside is that some functionality can only be accessed from the default home screen. However µLauncher could expose a simple API, like:
* choose an action: µLauncher shows the UI for choosing and returns a handle for the action selected by the user
* trigger a previously chosen action


Any opinions on this? Do you think this is an elegant minimalistic solution or just overly complex and unnecessarily complicated?
